shared/constants/types/fs.js

```diff
@@ -2,7 +2,27 @@
 import * as I from 'immutable'
 
 export type _State = {
-  counter: number,
+  path: Path,
+  pathItems: I.Map<Path, PathItem>,
 }
 
 export type State = I.RecordOf<_State>
+
+export type PathItem = {
```

Comment:
this type should be another record actually. so i'd rename this to _PathItem and have a type PathItem = I.RecordOf<_PathItem>.
Typically with immutable we want everything in the hierarchy to be immutable else you'll run into issue.

sidebar: one reason you want immutable with redux is that it helps ensure sub items that haven't changed will be === (their hash trie thing is pretty cool). We also want the opposite thing, which is to say if a prop in an object changes we want the parent object to actually change so mapStateToProps (etc) will equality fail and cause a rerender.

for example
const obj = {foo: 'bar'}
const obj2 = obj
obj.foo = 'whut'
obj === obj2 // whoops

So this can cause you to update the data and yet nothing renders. As a side effect its possible the map gets updated or something so you might be masking the issue but it really should be avoided.

The other issue is if you use immutable's helpers they mostly assume its a hierarchy of immutable objects. aka
myObject.getIn(['this', 'is', 'a path', 'of', 'properties','that','are','nested'])

so if these are all maps of maps of records of maps of records, etc it'll work fine. if its maps of objects of maps of records of objects it *could* fail.

easier if its all immutable top to bottom